fdbserver/GlobalTagThrottler.actor.cpp

```diff
@@ -0,0 +1,191 @@
+			}
+
+			// TODO: Implement expiration logic
+			if (!readLimit.present() && !writeLimit.present()) {
+				return {};
+			} else {
+				auto totalLimit = SERVER_KNOBS->GLOBAL_TAG_THROTTLING_MIN_RATE;
+				if (!readLimit.present()) {
+					totalLimit = std::max(totalLimit, writeLimit.get());
+				} else if (!writeLimit.present()) {
+					totalLimit = std::max(totalLimit, readLimit.get());
+				} else {
+					totalLimit = std::max(totalLimit, std::min(readLimit.get(), writeLimit.get()));
+				}
+				return ClientTagThrottleLimits(totalLimit, std::numeric_limits<double>::max());
+			}
+		}
+	};
+
+	Database db;
+	UID id;
+	std::map<TransactionTag, QuotaAndCounters> trackedTags;
+	uint64_t throttledTagChangeId{ 0 };
+
+	ACTOR static Future<Void> monitorThrottlingChanges(GlobalTagThrottlerImpl* self) {
+		loop {
+			state ReadYourWritesTransaction tr(self->db);
+
+			loop {
+				// TODO: Clean up quotas that have been removed
+				try {
+					tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+					tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+
+					state RangeResult currentQuotas = wait(tr.getRange(tagQuotaKeys, CLIENT_KNOBS->TOO_MANY));
+					TraceEvent("GlobalTagThrottler_ReadCurrentQuotas").detail("Size", currentQuotas.size());
+					for (auto const kv : currentQuotas) {
+						auto tag = kv.key.removePrefix(tagQuotaPrefix);
+						auto quota = ThrottleApi::TagQuotaValue::fromValue(kv.value);
+						self->trackedTags[tag].setQuota(quota);
+					}
+
+					++self->throttledTagChangeId;
+					// FIXME: Should wait on watch instead
+					// wait(tr.watch(tagThrottleSignalKey));
+					wait(delay(5.0));
+					TraceEvent("GlobalTagThrottler_ChangeSignaled");
+					TEST(true); // Global tag throttler detected quota changes
+					break;
+				} catch (Error& e) {
+					TraceEvent("GlobalTagThrottlerMonitoringChangesError", self->id).error(e);
+					wait(tr.onError(e));
+				}
+			}
+		}
+	}
+
+public:
+	GlobalTagThrottlerImpl(Database db, UID id) : db(db), id(id) {}
+	Future<Void> monitorThrottlingChanges() { return monitorThrottlingChanges(this); }
+	void addRequests(TransactionTag tag, int count) { trackedTags[tag].addTransactions(count); }
+	uint64_t getThrottledTagChangeId() const { return throttledTagChangeId; }
+	PrioritizedTransactionTagMap<ClientTagThrottleLimits> getClientRates() {
+		// TODO: For now, only enforce total throttling rates.
+		// We should use reserved quotas as well.
+		PrioritizedTransactionTagMap<ClientTagThrottleLimits> result;
+		for (const auto& [tag, quotaAndCounters] : trackedTags) {
+			// Currently there is no differentiation between batch priority and default priority transactions
+			auto const limit = quotaAndCounters.getTotalLimit();
+			if (limit.present()) {
+				result[TransactionPriority::BATCH][tag] = result[TransactionPriority::DEFAULT][tag] = limit.get();
+			}
+		}
+		return result;
+	}
```

Comment:
Here we use the total quota to calculate the total limit. Where will the reserved quota be used? I guess a future PR will add that part?

Reply:
Yeah, the reserved quota enforcement will go into a separate PR.